### PR TITLE
[https://nvbugs/6156492][fix] Fix disaggregated usage propagation

### PR DESCRIPTION
**Pull request by reasonsolo**
## Summary by CodeRabbit

* **New Features**
  * Added context response usage tracking and propagation across disaggregated serving phases.
  * Token usage now includes cached token accounting alongside prompt and completion tokens.
  * Context usage information is now carried through to generation response outputs.
  * Supports both streaming and non-streaming chat and completion endpoints.

* **Tests**
  * Updated existing tests to verify token usage data structure and propagation.
  * Added tests validating usage rewriting for generation postprocessors.

[Review Change Stack](https://app.coderabbit.ai/change-stack/NVIDIA/TensorRT-LLM/pull/14177)

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by fredricz-20070104 — APPROVED** [review]

**Review by lishicheng1996-nv — APPROVED** [review]

**Comment by lishicheng1996-nv** [comment]
It works great!

File: tests/integration/test_lists/qa/llm_function_rtx6k.txt [hunk 1]
```diff
@@ -224,21 +256,28 @@ test_e2e.py::test_ptp_quickstart_advanced[Llama3.1-8B-FP8-llama-3.1-model/Llama-
...
 test_e2e.py::test_eagle3_output_repetition_4gpus[Qwen3/saved_models_Qwen3-235B-A22B_nvfp4_hf-Qwen3/qwen3-235B-eagle3]
 
+
 disaggregated/test_disaggregated.py::test_disaggregated_single_gpu[TinyLlama-1.1B-Chat-v1.0]
 disaggregated/test_disaggregated.py::test_disaggregated_multi_gpu[TinyLlama-1.1B-Chat-v1.0]
 disaggregated/test_disaggregated.py::test_disaggregated_cuda_graph[TinyLlama-1.1B-Chat-v1.0]
 disaggregated/test_disaggregated.py::test_disaggregated_load_balance[TinyLlama-1.1B-Chat-v1.0]
 disaggregated/test_disaggregated.py::test_disaggregated_cache_aware_balance[TinyLlama-1.1B-Chat-v1.0]
 disaggregated/test_disaggregated.py::test_disaggregated_trtllm_sampler[TinyLlama-1.1B-Chat-v1.0]
+disaggregated/test_disaggregated.py::test_disaggregated_overlap[TinyLlama-1.1B-Chat-v1.0]
```
Comment (xinhe-nv) [inline tests/integration/test_lists/qa/llm_function_rtx6k.txt]: do we need to add the case into qa/llm_function_core.txt?

Comment (reasonsolo) [reply tests/integration/test_lists/qa/llm_function_rtx6k.txt]: It's not a new case, I just added a new check  at the end of the test. So this change is unnecessary.

**Review by syuoni — APPROVED** [review]
LLM API changes look good, thanks!

File: tensorrt_llm/_torch/disaggregation/native/auxiliary.py [hunk 2]
```diff
@@ -108,13 +115,17 @@ def __init__(self, max_slot_num: int, beam_width: int, max_draft_len: int, devic
         self._token_counts_buffer = torch.zeros(
             self._max_slot_num, 2, dtype=data_type, device=self._device
         )
+        self._usage_buffer = torch.zeros(
```
Comment (Shixiaowei02) [inline tensorrt_llm/_torch/disaggregation/native/auxiliary.py]:
```suggestion
        self._prompt_token_counts_buffer = torch.zeros(
```

File: tensorrt_llm/_torch/disaggregation/native/auxiliary.py [hunk 3]
```diff
@@ -204,6 +217,28 @@ def fill_slot(self, slot: int, request: LlmRequest) -> None:
                 [len(first_gen_tokens), len(draft_tokens)], dtype=torch.int32, device=self._device
             )
         )
+        prompt_tokens, cached_tokens = self._get_ctx_usage(request)
+        self._usage_buffer[slot].copy_(
+            torch.tensor([prompt_tokens, cached_tokens], dtype=torch.int32, device=self._device)
+        )
+
+    @staticmethod
+    def _get_ctx_usage(request: LlmRequest) -> tuple[int, int]:
```
Comment (Shixiaowei02) [inline tensorrt_llm/_torch/disaggregation/native/auxiliary.py]:
```suggestion
    def _resolve_prompt_token_counts(request: LlmRequest) -> tuple[int, int]:
```

File: tensorrt_llm/_torch/disaggregation/native/auxiliary.py [hunk 4]
```diff
@@ -204,6 +217,28 @@ def fill_slot(self, slot: int, request: LlmRequest) -> None:
...
+        ctx_response_usage = getattr(
```
Comment (Shixiaowei02) [inline tensorrt_llm/_torch/disaggregation/native/auxiliary.py]: Since `LlmRequest` is a concrete type, do we still need to use `getattr` here?

File: tensorrt_llm/_torch/disaggregation/native/auxiliary.py [hunk 5]
```diff
@@ -204,6 +217,28 @@ def fill_slot(self, slot: int, request: LlmRequest) -> None:
...
+            getattr(request, "py_disaggregated_params", None), "ctx_response_usage", None
```
Comment (Shixiaowei02) [inline tensorrt_llm/_torch/disaggregation/native/auxiliary.py]: Do we need to treat the upstream `ctx_response_usage` as authoritative here? If not, we might be able to simplify the logic.

File: tensorrt_llm/_torch/disaggregation/native/auxiliary.py [hunk 6]
```diff
@@ -213,3 +248,16 @@ def get_slot_tokens(self, slot: int) -> tuple[list[int], list[int]]:
         draft_tokens = self._draft_tokens_buffer[slot][:draft_len].tolist()
 
         return first_gen_tokens, draft_tokens
+
+    def get_slot_data(self, slot: int) -> tuple[list[int], list[int], dict[str, Any]]:
```
Comment (Shixiaowei02) [inline tensorrt_llm/_torch/disaggregation/native/auxiliary.py]: Could we lift the OpenAI usage dict assembly out of `AuxBuffer` and let the caller (`unpack_aux`, or alongside the new helpers in `disagg_utils.py`) shape it? Returning raw (`prompt_tokens`, `cached_tokens`) keeps the transport layer free of the OpenAI usage schema.

File: tensorrt_llm/disaggregated_params.py [hunk 7]
```diff
@@ -51,6 +53,7 @@ class DisaggregatedParams:
     ctx_dp_rank: Optional[int] = None
     ctx_info_endpoint: Optional[str] = None
     schedule_style: Optional[DisaggScheduleStyle] = None
+    ctx_response_usage: Optional[Dict[str, Any]] = None
```
Comment (Shixiaowei02) [inline tensorrt_llm/disaggregated_params.py]:
```suggestion
    ctx_usage: Optional[Dict[str, Any]] = None
```

Comment (reasonsolo) [reply tensorrt_llm/disaggregated_params.py]: Thanks, all of your comments has been addressed. Please review again. Thanks!

**Comment by pcastonguay** [comment]
@Shixiaowei02 could you have another look when you have a chance? Thanks.

**Review by Shixiaowei02 — APPROVED** [review]
